File: java/data_structures_examples/BinSearch.java
package data_structures_examples;

public class BinSearch {
    // https://leetcode.com/problems/search-insert-position/description/
    static class Solution1 {
        public int searchInsert1(int[] nums, int target) {
            var lo = -1;
            var hi = nums.length;

            while (hi - lo > 1) {
                var mid = (lo + hi) / 2;

                if (nums[mid] < target) {
                    lo = mid;
                } else {
                    hi = mid;
                }
            }
            return hi;
        }

        public int searchInsert2(int[] nums, int target) {
            var lo = 0;
            var hi = nums.length;
            while (lo < hi) {
                var mid = (lo + hi) / 2;
                if (nums[mid] < target) {
                    lo = mid + 1;
                } else {
                    hi = mid;
                }
            }
            return lo;
        }

        public int searchInsert3(int[] nums, int target) {
            var lo = 0;
            var hi = nums.length - 1;
            while (lo <= hi) {
                var mid = (lo + hi) / 2;
                if (nums[mid] == target) {
                    return mid;
                }
                if (nums[mid] < target) {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }
            return lo;
        }
    }
}

File: java/data_structures_examples/bin_search.rs
// https://leetcode.com/problems/element-appearing-more-than-25-in-sorted-array/description/
pub fn find_special_integer_3(arr: Vec<i32>) -> i32 {
    fn upper_bound(arr: &Vec<i32>, target: i32) -> i32 {
        let mut lo = 0;
        let mut hi = arr.len();

        while lo < hi {
            let mid = lo + (hi - lo) / 2;
            if arr[mid] > target {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        lo as i32
    }

    fn lower_bound(arr: &Vec<i32>, target: i32) -> i32 {
        let mut lo = 0;
        let mut hi = arr.len();

        while lo < hi {
            let mid = lo + (hi - lo) / 2;
            if arr[mid] >= target {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        lo as i32
    }

    let n = arr.len();
    let candidates = vec![arr[n / 4], arr[n / 2], arr[3 * n / 4]];
    let target = n / 4;

    for c in candidates {
        let left = lower_bound(&arr, c);
        let right = upper_bound(&arr, c) - 1;

        if right - left + 1 > target as i32 {
            return c;
        }
    }

    -1
}
